Refactor General Ledger

jniles opened the pull request:
This PR refactors the General Ledger to make it more useful from an accounting perspective.

**Code Refactor**
 1. The server-side SQL query has been rewritten to be more performant and easier to read.  In particular, the WHERE query filters on fiscal year instead of first looking up the periods in the fiscal year.
 2. Removed the Account Slip code - it was essentially a duplication of effort from the Account Statement code.
 3. Renamed all variables references `general-ledger-accounts` to `general-ledger`.  We had folded the previous "General Ledger" into the Journal, so only this one remains.

**Features**
 1. A `<select>` element to the grid footer to use to switch the fiscal year and removes the Fiscal Year modal.  This was requested by @WayneNiles and improves the UX by allowing users to switch fiscal years on the fly.  By removing the modal, we also alleviate the maintenance burden on developers.
 2. A "check row" at the bottom of the General Ledger has been implemented, showing the sum of each column with period balances.  _These columns should always sum to zero_.  If not, it means that there is an imbalanced transaction somewhere.
 3. Title accounts are now show and hold the sum of the balances of their children.  This allows a really quick comparison of bulk amounts, such as Income versus Expense.  This has been implemented using a `Tree` class, available to be used in other reports.
 4. Added title account highlighting and indentation to both the client and server-side generated reports for the General Ledger.  The indentation is set to `15px` by default.
 5. Add a link to the Account Statement module.  This allows a user to directly inspect the values from the General Ledger in the Account Statement.

**Screenshots**
![newgl](https://user-images.githubusercontent.com/896472/35456000-5a04432a-02d4-11e8-812d-760446e3a6ae.PNG)
_Fig 1: New General Ledger Client_

![newglreport](https://user-images.githubusercontent.com/896472/35456057-8e33d76e-02d4-11e8-8418-db15de5c8195.PNG)
_Fig 2: New General Ledger Report w/ Indentation_


 Closes #1706.  Closes #2459.  Closes #2463. Closes #2373.

jniles committed:
perf(general_leger): optimize GL Matrix

This commit optimizes the General Ledger matrix query by only doing a
single SELECT with a GROUP BY.  It also removes the need for
pre-querying period ids and uses the raw fiscal year id instead.  This
is the first step to further optimizations of calculating the balances
of title accounts by summing down the account tree.

jniles committed:
refactor(general ledger): <select> FY in footer

This commit adds a footer to the General Ledger with the Fiscal Year in
it.  This allows rapid changes to the view without need to have multiple
clicks.

jniles committed:
feat(Tree): implement tree structure w/ tests

This commit implements a Tree class on the server for constructing trees
out of accounts, units, or other tree structures.  The only requirement
is that the objects have references pointing them to their parents.

jniles committed:
feat(Tree): implement sumOnProperty()

This commit implements the sumOnProperty() function to sum tree nodes
on a given property.  Parent nodes are given the values of the sums of
their children.

jniles committed:
feat(General Ledger): implement title summing

This commit implements basic summing on the general ledger's title
accounts.  The summing is performed on the server in JavaScript, but
could be done on the client just as easily.

jniles committed:
feat(General Ledger): prune empty accounts

This commit makes sure empty accounts are pruned from the General Ledger
report.

mbayopanda commented on client/src/modules/general-ledger/general-ledger-accounts.ctrl.js:
```diff
@@ -212,52 +235,27 @@ function GeneralLedgerAccountsController(
-    account.isTitleAccount = account.type_id === bhConstants.accounts.TITLE;
+  function removeZeroes(account) {
+    // remove zero values from the matrix to render as empty cells.
+    fields.forEach(function (field) {
```
👍

mbayopanda commented on server/lib/Tree.js:
```diff
@@ -0,0 +1,138 @@
+/**
+ * @class Tree
```
👍 awesome... Good job

mbayopanda commented on test/server-unit/Tree.spec.js:
```diff
@@ -0,0 +1,119 @@
+const Tree = require('../../server/lib/Tree');
+const { expect } = require('chai');
+
+function TreeUnitTests() {
```
💃 Excellent

mbayopanda reviewed (commented):
LGTM

I just noticed that there is an error when click on `action -> Extrait de compte` to get the `extrait de compte` report.

jniles committed:
feat(General Ledger): use Account Statement Report

This commit removes the old Account Slip report link and replaces it
with the account statement report link from the General Ledger.

Closes #1706.

jniles committed:
chore(General Ledger): remove accountSlip

This commit removes the account slip which is no longer used.  It also
adds an indent to the account list so that nested items are clearly
below their parent.  Finally, title accounts are rendered in bold.

jniles committed:
chore(General Ledger): remove accounts from name

This commit renames the general ledger files to remove the "accounts"
portion of the name from all identifiers.  This is the real General
Ledger, since the previous version was folded into the Journal.

jniles committed:
chore(General Ledger): indent accounts

This commit indents the accounts in the tree to emphasize their
relationships.

jniles committed:
feat(General Ledger): implement checksum footer

This commit implements a check footer at the bottom of the grid that
sums the values in the columns.  As a rule of accounting, summing the
chart of accounts should give a zero.  Thus, any columns which do not
sum to 0 denote a problem in the dataset.

Closes #2463.

jniles committed:
feat(General Ledger): link Account Statement

This commit links the General Ledger to the Account Statement via the
dropdown menu in the General Ledger.

Closes #2459.

mbayopanda reviewed (approved):
It works very well. :+1
